framework/src/modules/chain/submodules/blocks/process.js

```diff
@@ -427,11 +435,242 @@ class Process {
+				block.height
+			)} slot: ${slots.getSlotNumber(block.timestamp)} generator: ${
+				block.generatorPublicKey
+			}`
+		);
+	}
+
+	/**
+	 * Block IDs are the same ~ Blocks are equal
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleSameBlockReceived(block) {
+		library.logger.debug('Block already processed', block.id);
+	}
+
+	/**
+	 * Block received is correct
+	 * @param block
+	 * @returns {Promise}
+	 * @private
+	 */
+	_handleGoodBlock(block) {
+		return this._processBlock(block);
+	}
+
+	/**
+	 * Double forging. Last block stands
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleDoubleForging(block) {
+		library.logger.warn(
+			'Delegate forging on multiple nodes',
+			block.generatorPublicKey
+		);
+		library.logger.info('Last block stands');
+		// TODO: Implement Proof of Misbehavior
+	}
+
+	/**
+	 * Tie break: two competing blocks by different delegates at the same height.
+	 * @param lastBlock
+	 * @param newBlock
+	 * @returns {Promise}
+	 * @private
+	 */
+	async _handleDoubleForgingTieBreak(newBlock, lastBlock) {
+		try {
+			// Normalize and check schema
+			const normalizedNewBlock = library.logic.block.objectNormalize(
+				_.cloneDeep(newBlock)
+			);
+
+			// Validate block has been generated by correct delegate
+			await this._validateGenerator(normalizedNewBlock, lastBlock);
+
+			// Verify block data, signatures, etc
+			// TODO: Maybe `verifyReceipt` could throw errors instead of returning checks.
+			const check = modules.blocks.verify.verifyReceipt(normalizedNewBlock);
+
+			if (!check.verified) {
+				library.logger.error(
+					`Block ${normalizedNewBlock.id} verification failed`,
+					check.errors.join(', ')
+				);
+				// Return first error from checks
+				throw check.errors[0];
+			}
+		} catch (error) {
+			// Should there be any errors while validating and verifying,
+			// last block is kept as the tip of the chain
+			library.logger.error(
+				'Case 4 fork recovery failed',
+				convertErrorsToString(error)
+			);
+			throw error;
+		}
+
+		// If the new block is correctly validated and verified,
+		// last bloc kis deleted and new block is added to the tip chain
+		library.logger.info('Deleting last block due to case 4');
+		await deleteLastBlock();
+
+		return this._processBlock(newBlock);
+	}
+
+	/**
+	 * Move to a different chain
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleMovingToDifferentChain() {
+		// TODO: Move to a different chain.
+	}
+
+	/**
+	 * Update last receipt and process newly received block.
+	 * @param block
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_processBlock(block) {
+		return promisify(__private.receiveBlock)(block); // TODO: Convert __private.receiveBlock to async, move implementation here to async and remove from __private.
```

Comment:
Why to promising every time? Please create a proxy function `__private.receiveBlockPromisify` and use it here.